File: backend/app/simulation/traffic_lights.py
import enum
from datetime import datetime

from app.db.models.traffic_light import TrafficLight


class LightState(enum.StrEnum):
    RED = "red"
    YELLOW = "yellow"
    GREEN = "green"


def light_state_at(light: TrafficLight, now: datetime) -> LightState:
    """Pure function of wall-clock time — no DB writes needed per tick, same
    principle already used for vehicle/truck route position (Этап 7/8).
    """
    cycle_length = light.red_seconds + light.yellow_seconds + light.green_seconds
    if cycle_length <= 0:
        return LightState.GREEN

    phase = (now.timestamp() + light.offset_seconds) % cycle_length
    if phase < light.red_seconds:
        return LightState.RED
    if phase < light.red_seconds + light.yellow_seconds:
        return LightState.YELLOW
    return LightState.GREEN

File: frontend/src/map/trafficLights.ts
import type { TrafficLight } from '../api/map'

export type LightState = 'red' | 'yellow' | 'green'

export function lightStateAt(light: TrafficLight, nowMs: number): LightState {
  const cycleLength = light.red_seconds + light.yellow_seconds + light.green_seconds
  if (cycleLength <= 0) return 'green'

  const nowSeconds = nowMs / 1000
  const phase = ((nowSeconds + light.offset_seconds) % cycleLength + cycleLength) % cycleLength

  if (phase < light.red_seconds) return 'red'
  if (phase < light.red_seconds + light.yellow_seconds) return 'yellow'
  return 'green'
}

export const LIGHT_STATE_COLORS: Record<LightState, string> = {
  red: '#e53935',
  yellow: '#fdd835',
  green: '#43a047',
}
